### PR TITLE
webui: Fix selecting generated output issues during active streaming

### LINKED ISSUE #17132: Misc. bug: Webui: Problems selecting text while generating
### Name and Version

build: 6951 (9aa63374f) with cc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0 for x86_64-linux-gnu

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell

```

### Problem description & steps to reproduce

*Motivation*: While generating i read the text, and may want to select text parts and copy paste it to my writing text box (next prompt) - so I can adress specific parts of the answer. Then continue reading.

*Problem*: Selecting doesn't work properly while generating (so the by mouse selected text isn't properly selected). Not properly meaning: The text also above the selection is selected and the whole selection is dismissed after releasing the mouseclick.

### First Bad Commit

_No response_

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by allozaur**
Close #17132

- Implements incremental rendering: Splits markdown content into stable blocks and a single unstable block (initial draft created by @ServeurpersoCom):
  - Stable blocks (all but the last) are cached and only rendered once
  - Only the last block (unstable) is re-rendered during streaming
  - Prevents DOM reconstruction of already-rendered content, enabling smooth text selection
  - Using HAST node positions for stable block identification
- Reduces unnecessary re-processing of unchanged markdown blocks
- Improves code organization in `MarkdownContent` component

https://github.com/user-attachments/assets/ce669e68-4d4f-42fe-8ba3-4bb6e7e7d794

### CODE REVIEW

**Review by allozaur — COMMENTED** [review]
@ggerganov @ngxson @ServeurpersoCom

Please do some testing on your end as well and let me know if we aren't missing anything in this PR to address the issue.

**Comment by ServeurpersoCom** [comment]
Fantastic! I absolutely must stress test it @ MoE A3B on 5090 because my draft kept crashing after a while!

Edit : We have the same edge case bug as in the POC/Draft (video watched together). We need to try to narrow down by generating the content that causes it.

**Comment by ServeurpersoCom** [comment]
Running GPT-OSS-20B at 330 tok/s inference was faster than rendering, making it easier to trigger the race condition between stable/unstable block updates, solution: await tick() to force DOM sync
I can no longer reproduce the bug with this patch
I haven't noticed any performance drop, although it could be improved by limiting rendering to one per requestAnimationFrame or a submultiple of them later.

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte [hunk 1]
```diff
@@ -266,6 +339,81 @@
...
+
+		let unstableHtml = '';
+
+		if (processedChildren.length > stableCount) {
+			const unstableChild = processedChildren[stableCount];
+			unstableHtml = stringifyProcessedNode(processorInstance, processedRoot, unstableChild);
+		}
+
+		renderedBlocks = nextBlocks;
+		unstableBlockHtml = unstableHtml;
```
Comment (ServeurpersoCom) [inline tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte:273]:
```suggestion
		renderedBlocks = nextBlocks;
		await tick(); // Force DOM sync before updating unstable HTML block
		unstableBlockHtml = unstableHtml;
```

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte [hunk 2]
```diff
@@ -7,49 +7,43 @@
 	import remarkRehype from 'remark-rehype';
 	import rehypeKatex from 'rehype-katex';
 	import rehypeStringify from 'rehype-stringify';
-	import { copyCodeToClipboard, preprocessLaTeX } from '$lib/utils';
-	import { rehypeRestoreTableHtml } from '$lib/markdown/table-html-restorer';
+	import type { Root as HastRoot, RootContent as HastRootContent } from 'hast';
+	import type { Root as MdastRoot } from 'mdast';
 	import { browser } from '$app/environment';
+	import { onDestroy } from 'svelte';
```
Comment (ServeurpersoCom) [inline tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]:
```suggestion
	import { onDestroy, tick } from 'svelte';
```

**Comment by ggerganov** [comment]
Generally this also works on my end, but I do see occasionally bigger selections than expected. Maybe related to the race that @ServeurpersoCom found:


https://github.com/user-attachments/assets/a3965697-e567-41f6-a09c-aca8cbe3b0d8

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte [hunk 3]
```diff
@@ -7,49 +7,43 @@
...
-		document.head.appendChild(style);
-	}
-
-	$effect(() => {
-		const currentMode = mode.current;
-		const isDark = currentMode === 'dark';
-
-		loadHighlightTheme(isDark);
-	});
+	const themeStyleId = `highlight-theme-${uuid()}`;
```
Comment (ngxson) [inline tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: maybe we can replace `uuid()` with a simple global counter `idxThemeStyle`? (this should improve the performance a little bit if it is called repeatedly during rendering)

Something like:

```suggestion
	const themeStyleId = `highlight-theme-${window.idxThemeStyle++}`;
```

**Review by ggerganov — APPROVED** [review]
Actually, I did some more testing and the problem that I observed occurs only when I am trying to select text inside a code block that is currently being generated. After it gets closed, then selecting for that block works ok.

I think this is acceptable.

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte [hunk 4]
```diff
@@ -287,40 +436,99 @@
...
+		processorInstance: ReturnType<typeof processor>,
+		processedRoot: HastRoot,
+		child: unknown
+	) {
+		const root: HastRoot = {
+			...(processedRoot as HastRoot),
+			children: [child as never]
+		};
+
+		const html = processorInstance.stringify(root);
```
Comment (ngxson) [inline tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: probably this can be improved by removing the `stringify()` and instead, making `enhanceCodeBlocks` and `enchance*` transformations to travel the virtual DOM tree (aka `HastRoot`) and directly modify the tree

this should be more performant as we no longer have to stringify then parse it again inside each transformation (via `createElement`)

but this can be done in a folow-up PR if needed

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte [hunk 5]
```diff
@@ -64,136 +58,174 @@
...
+					if (className.startsWith('language-')) {
+						language = className.replace('language-', '');
+						break;
+					}
+				}
 
-			let language = 'text';
-			const classList = Array.from(codeElement.classList);
+				const rawCode = codeElement.textContent || '';
+				const codeId = `code-${uuid()}-${index}`;
```
Comment (ngxson) [inline tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: same remark here for `uuid()`

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte (hunk 3 above)
Comment (allozaur) [reply tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: Yeah, we could do that

Comment (ngxson) [reply tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: (Btw i think it's better to wrap the increment inside a function: getId(): number)

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte (hunk 4 above)
Comment (allozaur) [reply tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: In 1ce3baa9b7494555a8b97ebb9201eef0145db106 I've refactored this to use rehype plugins that operate directly on the HAST tree:

- Created `rehypeEnhanceLinks` - modifies link nodes in HAST
- Created `rehypeEnhanceCodeBlocks` - wraps code blocks in HAST

So before we had `HAST → stringify → createElement (parse) → modify DOM → innerHTML → createElement → modify → innerHTML` and now it's simplified to `HAST → modify HAST → modify HAST → stringify`

The final `stringify()` is still needed since Svelte renders via {@html}, but we eliminated the intermediate parse/stringify cycles that createElement + innerHTML were doing. This reduces the operation from 3 stringify + 2 parse calls down to just 1 stringify.

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte (hunk 3 above)
Comment (allozaur) [reply tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: Replaced UUID here with global id + counter variables with a10adb5674bc0f67bf1ef2862386b221b1397120

File: tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte (hunk 5 above)
Comment (allozaur) [reply tools/server/webui/src/lib/components/app/misc/MarkdownContent.svelte]: a10adb5674bc0f67bf1ef2862386b221b1397120

**Comment by allozaur** [comment]
> Actually, I did some more testing and the problem that I observed occurs only when I am trying to select text inside a code block that is currently being generated. After it gets closed, then selecting for that block works ok.
> 
> I think this is acceptable.

@ggerganov @ServeurpersoCom I've added some changes after @ngxson's review. Please re-test this on your ends.

**Comment by ServeurpersoCom** [comment]
No regression on my end: >10 long rich markdown generations with success, whereas the corruption was systematically occurring after 2 or 3 generations
Testing : GPT-OSS-20B
Prompt : Writes a long and rich markdown
+ Smartphone test OK

**Comment by ggerganov** [comment]
Found a bug introduced here - when you "Copy code", it forgets the whitespaces:

image

**Review by ngxson — APPROVED** [review]
Tested on my side, it works except for inside a generating code block as Georgi spotted earlier.

We can improve this in the future by somehow prevent setting innerHTML via `{@html block.html}`. Instead, I think it's best to have a system where it can take a `HastRoot` and use depth-first search to get the diff between 2 virtual DOM, and only update changed nodes on HTML. Probably there are some libraries already do all of these heavy-lifting works for us, but we can have a look later.

For now, I think this PR is good to merge.

**Comment by ServeurpersoCom** [comment]
Same on Windows, copy paste -> no `\n` or no `\r\n`

**Comment by allozaur** [comment]
> Found a bug introduced here - when you "Copy code", it forgets the whitespaces:
> image

@ggerganov @ServeurpersoCom this should be fixed with 30c2c18f24368164cf61d37edeeba10e2362442d

**Review by ggerganov — APPROVED** [review]

**Comment by thomasjfox** [comment]
Thanks so much for this one! :partying_face:

It fixes a usability annoyance for good. The users will love it.

**Comment by allozaur** [comment]
> Thanks so much for this one! :partying_face:
> 
> 
> 
> It fixes a usability annoyance for good. The users will love it.

Great to hear! It's still not in a perfect state and we are planning a better strategy for rendering the generated content, but it solves the most pressing issue.
